kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/commit/Gen2Committer.scala

```diff
@@ -0,0 +1,133 @@
+import io.lenses.streamreactor.connect.aws.s3.storage.Storage
+import org.apache.kafka.connect.errors.ConnectException
+
+import scala.io.Source
+import scala.util.control.NonFatal
+
+class Gen2Committer(storage: Storage,
+                    fileNamingStrategyFn: Topic => S3FileNamingStrategy,
+                    bucketAndPrefixFn: Topic => BucketAndPrefix) extends Committer {
+
+  private val logger = org.slf4j.LoggerFactory.getLogger(getClass.getName)
+
+  private val indexNamingStrategy = new S3IndexNamingStrategy
+
+  /**
+   * Commit records to S3 bucket.
+   *
+   * @param bucketAndPrefix      the bucket.
+   * @param topicPartitionOffset the offset.
+   * @param partitionValues      the partition values.
+   */
+  override def commit(bucketAndPrefix: BucketAndPrefix,
+                      topicPartitionOffset: TopicPartitionOffset,
+                      partitionValues: Map[PartitionField, String]): Unit = {
+    val fileNamingStrategy = fileNamingStrategyFn(topicPartitionOffset.topic)
+    val originalFilename = fileNamingStrategy.stagingFilename(
+      bucketAndPrefix,
+      topicPartitionOffset.toTopicPartition,
+      partitionValues
+    )
+    val finalFilename = fileNamingStrategy.finalFilename(
+      bucketAndPrefix,
+      topicPartitionOffset,
+      partitionValues
+    )
+
+    val topicPartition = topicPartitionOffset.toTopicPartition
+    val currentIndex = findIndexPath(topicPartition)
+    val newIndex = indexNamingStrategy.indexFilename(BucketAndPrefix(bucketAndPrefix.bucket), topicPartitionOffset)
+
+    // create new index
+    storage.putBlob(newIndex, finalFilename.path)
+    // commit
+    storage.rename(originalFilename, finalFilename)
+    // remove current index
+    currentIndex.foreach(indexPath => storage.remove(indexPath))
+  }
+
+  /**
+   * Find latest offset per partition.
+   *
+   * @param partitions the set of topic partitions
+   * @return the offset per partition <code>Map</code>.
+   */
+  override def latest(partitions: Set[TopicPartition]): Map[TopicPartition, Offset] = {
+    partitions.collect {
+      case topicPartition: TopicPartition =>
+        findIndexPath(topicPartition)
+          .flatMap(bucketAndPath => IndexFileName.from(bucketAndPath.path, indexNamingStrategy))
+          .map(file => TopicPartitionOffset(file.topic, file.partition, file.offset)) match {
+          case Some(topicPartitionOffset) => Some(topicPartition, topicPartitionOffset.offset)
+          case None => None
+        }
+    }.flatten.toMap
+  }
+
+  /**
+   * Find latest index path for a partition.
+   *
+   * In case multiple indexes are available resolves conflict
+   * by removing the non-committed index.
+   *
+   * @param topicPartition the topic partition
+   * @return the bucket and path of latest index.
+   */
+  private def findIndexPath(topicPartition: TopicPartition): Option[BucketAndPath] = {
```

Comment:
Style comment. I would split this function. it does more than just find. otherwise maybe rename it to point out it is doing more